packages/medusa/src/services/discount-condition.ts

```diff
@@ -0,0 +1,237 @@
+import { MedusaError } from "medusa-core-utils"
+import { BaseService } from "medusa-interfaces"
+import { EntityManager } from "typeorm"
+import { EventBusService } from "."
+import { Discount, DiscountCondition, DiscountConditionType } from "../models"
+import { DiscountRepository } from "../repositories/discount"
+import { DiscountConditionRepository } from "../repositories/discount-condition"
+import { DiscountRuleRepository } from "../repositories/discount-rule"
+import { UpsertDiscountConditionInput } from "../types/discount"
+import { PostgresError } from "../utils/exception-formatter"
+
+/**
+ * Provides layer to manipulate discount conditions.
+ * @implements {BaseService}
+ */
+class DiscountConditionService extends BaseService {
+  protected readonly manager_: EntityManager
+  protected readonly discountRuleRepository_: typeof DiscountRuleRepository
+  protected readonly discountRepository_: typeof DiscountRepository
+  protected readonly discountConditionRepository_: typeof DiscountConditionRepository
+  protected readonly eventBus_: EventBusService
+
+  constructor({
+    manager,
+    discountRuleRepository,
+    discountRepository,
+    discountConditionRepository,
+    eventBusService,
+  }) {
+    super()
+
+    /** @private @const {EntityManager} */
+    this.manager_ = manager
+
+    /** @private @const {DiscountRuleRepository} */
+    this.discountRuleRepository_ = discountRuleRepository
+
+    /** @private @const {DiscountRepository} */
+    this.discountRepository_ = discountRepository
+
+    /** @private @const {DiscountConditionRepository} */
+    this.discountConditionRepository_ = discountConditionRepository
+
+    /** @private @const {EventBus} */
+    this.eventBus_ = eventBusService
+  }
+
+  withTransaction(transactionManager: EntityManager): DiscountConditionService {
+    if (!transactionManager) {
+      return this
+    }
+
+    const cloned = new DiscountConditionService({
+      manager: transactionManager,
+      discountRuleRepository: this.discountRuleRepository_,
+      discountRepository: this.discountRepository_,
+      discountConditionRepository: this.discountConditionRepository_,
+      eventBusService: this.eventBus_,
+    })
+
+    cloned.transactionManager_ = transactionManager
+
+    return cloned
+  }
+
+  async retrieve(
+    conditionId: string,
+    discountId: string
+  ): Promise<(DiscountCondition & { discount: Discount }) | never> {
```

Comment:
thought: this couples conditions and discounts quite tightly; say you are building a script or some sort of other custom functionality that does operations on the condition; in that case, it is a bit annoying to have to know the condition and discount ids in order to just retrieve the condition data. 

I can see why it has been included (to ensure that discount + condition combination is valid), but I don't think it is the ConditionService's responsibility to enforce that.

Reply:
@srindom i would like to introduce something I mentioned which os the middleware/guard/pipe approches which would allow us to decoupled those kind of check out and compose our api with more flexibility and ensure the coherence before hitting the endpoints. This is something i will write up soon. At the moment, we can either use the repo in the endpoint just to join and ensure that the condition belongs to the discount, or call the discountService in the endpoint and adding the condition relation for the check purpose, or split that method and add a new retrieveWithDiscount which would take care to call the retrieve and attach the discount or throw. What do you think?

Reply:
@adrien2p @srindom This is somewhat similar to what @pKorsholm was adhering to, but just in `upsertConditions`. I am up for making the DiscountConditionService more dumb by 
- moving the relational check to the controller
- refactoring `upsertConditions` to only take in data and not be concerned with what is then provided (this is what @pKorsholm and I discussed previously)
- refactoring `retrieve` to only use condition id and presume that the relational check is done elsewhere